Feat/node insertion

DieracDelta opened the pull request:
This is a draft PR for AST node insertion. Objectives:

- [x] create AST nodes for attribute sets
- [x] convert between `Input` type and attribute set
- [ ] clean code 
  - [ ] so many places need `&str` instead of `String` 
  - [ ] better file naming to match rest of project
  - [ ] clippy
- [ ] tie into main control loop so we can create nodes
- [ ] more tests

zseri commented on src/parser/input_utils.rs:
```diff
@@ -0,0 +1,98 @@
+    GreenNode::new(kind, children)
+}
+
+pub fn gen_attr_set(attr_pairs: Vec<(String, String)>) -> GreenNode{
```
```suggestion
pub fn gen_attr_set(attr_pairs: &[(String, String)]) -> GreenNode{
```

DieracDelta replied:
I'm trying to do this, and running into issues (here)[https://github.com/DieracDelta/flake_generator/blob/feat/node_insertion/src/parser/input_utils.rs#L163]. It looks like there's not an easy way to convert from a vector to an array..

zseri commented on src/parser/input_utils.rs:
```diff
@@ -0,0 +1,98 @@
+pub fn gen_attr_set(attr_pairs: Vec<(String, String)>) -> GreenNode{
+    let new_attr_pairs : Vec<(GreenNode, GreenNode)> =
+        attr_pairs.iter().map(|(key, value)| (new_key(key.to_string()), new_string(value.to_string()))).collect();
+    new_attr_set(new_attr_pairs)
```
from a performance standpoint it is a better idea to just merge `gen_attr_set` and `new_attr_set`, to avoid the allocation of `new_attr_pairs`, which then get `.clone()`d directly again, which should be avoided.

DieracDelta replied:
What I ended up doing here is implementing the from trait so I can just convert directly from an Input to a GreenNode. The `gen_*` functions can be deleted.

DieracDelta commented on src/parser/input_utils.rs:
```diff
@@ -0,0 +1,98 @@
+    new_key_value(key_node, value_node)
+}
+
+pub fn new_key_value(key : GreenNode, value: GreenNode) -> GreenNode {
```
The interface isn't clean. I agree with you--I'll get rid of the `new_*`.

DieracDelta commented:
Input generation works, and I've added in a nixpkgs_fmt pass to make sure things look nice. I've added a couple of tests as well. Code quality is poor; lots of unused imports, dead code, etc. A bunch of the processing done in main.rs does not belong. Inputs_utils needs to be split out into something more modular that matches with the rest of the style. Not ready to merge, but getting there.

DieracDelta commented:
Also, inputs need to be (1) checked, and (2) there should be some sort of type generation. Ideally, we should support other types of  flake inputs that covers the entire spec. Path/git/, enabling submodules, revisions, etc. Essentially, everything specified [here](https://gist.github.com/edolstra/40da6e3a4d4ee8fd019395365e0772e). I think I'd prefer to do these in separate PRs though. The scaffolding here can define the basis for further work.

zseri commented on src/main.rs:
```diff
@@ -1,9 +1,15 @@
-use parser::utils::remove_input;
+use parser::input_utils::{merge_attr_sets, Input};
+use parser::utils::{get_node_idx, remove_input, search_for_attr, NixNode};
+use rowan::GreenNode;
+use rowan::NodeOrToken;
+use rowan::SyntaxNode;
```
```suggestion
use rowan::{GreenNode, NodeOrToken, SyntaxNode};
```

zseri commented on src/main.rs:
```diff
@@ -21,6 +27,12 @@ impl ActionStack {
     }
 
+    fn push_seq(&mut self, actions: impl IntoIterator<Item = UserAction>) {
+        actions
+            .into_iter()
+            .for_each(|action| self.inner.push(action))
```
```suggestion
        self.inner.extend(actions.into_iter());
```

zseri commented on src/main.rs:
```diff
@@ -64,49 +76,109 @@ fn main() {
+                    let mut i = user_data
+                        .new_input
+                        .as_ref()
+                        .and_then(|x: &Input| Some(x.clone()))
```
```suggestion
                        .cloned()
```

zseri commented on src/main.rs:
```diff
@@ -64,49 +76,109 @@ fn main() {
+                    let new_input: GreenNode = user_data
+                        .new_input
+                        .as_ref()
+                        .and_then(|x| Some(x.clone()))
```
```suggestion
                        .cloned()
```

zseri commented on src/parser/input_utils.rs:
```diff
@@ -0,0 +1,166 @@
+}
+
+pub fn new_key_value(key: GreenNode, value: GreenNode) -> GreenNode {
+    let kind = NixLanguage::kind_to_raw(NODE_KEY_VALUE);
+    let assign_kind = NixLanguage::kind_to_raw(TOKEN_ASSIGN);
+    let whitespace_kind = NixLanguage::kind_to_raw(TOKEN_WHITESPACE);
+    let semicolon_kind = NixLanguage::kind_to_raw(TOKEN_SEMICOLON);
+    let children = vec![
+        NodeOrToken::Node(key),
+        NodeOrToken::Token(GreenToken::new(whitespace_kind, " ")),
+        NodeOrToken::Token(GreenToken::new(assign_kind, "=")),
+        NodeOrToken::Token(GreenToken::new(whitespace_kind, " ")),
+        NodeOrToken::Node(value),
+        NodeOrToken::Token(GreenToken::new(semicolon_kind, ";")),
+        NodeOrToken::Token(GreenToken::new(whitespace_kind, "\n")),
+    ];
```
```suggestion
    use NixLanguage::kind_to_raw;
    use NodeOrToken::{Node, Token};
    let kind = kind_to_raw(NODE_KEY_VALUE);
    let assign_kind = kind_to_raw(TOKEN_ASSIGN);
    let whitespace_kind = kind_to_raw(TOKEN_WHITESPACE);
    let semicolon_kind = kind_to_raw(TOKEN_SEMICOLON);
    let children = vec![
        Node(key),
        Token(GreenToken::new(whitespace_kind, " ")),
        Token(GreenToken::new(assign_kind, "=")),
        Token(GreenToken::new(whitespace_kind, " ")),
        Node(value),
        Token(GreenToken::new(semicolon_kind, ";")),
        Token(GreenToken::new(whitespace_kind, "\n")),
    ];
```

DieracDelta replied:
Can't import NixLanguage::kind_to_raw. Not sure why....

zseri replied:
Oh... (that might be the case because the method is defined in a trait impl)
Maybe create a small wrapper for it then:
```rust
#[inline(always)]
fn kind_to_raw(x: rnix::SyntaxKind) -> rowan::SyntaxKind {
    NixLanguage::kind_to_raw(x)
}
```

zseri commented on src/parser/input_utils.rs:
```diff
@@ -0,0 +1,166 @@
+        .children()
+        .into_iter()
+        .filter(|node| node.kind() == NixLanguage::kind_to_raw(NODE_KEY_VALUE))
+        .map(|x| match x {
+            NodeOrToken::Node(x) => NodeOrToken::Node(x.clone()),
+            NodeOrToken::Token(x) => NodeOrToken::Token(x.clone()),
+        })
```
```suggestion
        .clone()
```

DieracDelta replied:
`.clone` isn't working here. I'm getting a  type error; for some reason .clone() turns the type into a NodeOrToken<&Node, &Token>, which needs to be dereferenced.

zseri commented on src/parser/input_utils.rs:
```diff
@@ -0,0 +1,166 @@
+// TODO give all the tokens their own constructors
+/// inputs are: [(lhs, rhs), ...]
+pub fn new_attr_set(attr_pairs: Vec<(GreenNode, GreenNode)>) -> GreenNode {
+    let pairs: Vec<NodeOrToken<_, _>> = attr_pairs
+        .iter()
+        .map(move |(k, v)| NodeOrToken::Node(new_key_value(k.clone(), v.clone())))
+        .collect::<Vec<_>>();
+    let open_curly_kind = NixLanguage::kind_to_raw(TOKEN_CURLY_B_OPEN);
+    let close_curly_kind = NixLanguage::kind_to_raw(TOKEN_CURLY_B_CLOSE);
+    let attr_set_kind = NixLanguage::kind_to_raw(NODE_ATTR_SET);
+    let whitespace_kind = NixLanguage::kind_to_raw(TOKEN_WHITESPACE);
+    let mut token_vec = Vec::new();
+    token_vec.push(vec![NodeOrToken::Token(GreenToken::new(
+        open_curly_kind,
+        "{",
+    ))]);
+    token_vec.push(vec![NodeOrToken::Token(GreenToken::new(
+        whitespace_kind,
+        "\n",
+    ))]);
+    token_vec.push(pairs);
+    token_vec.push(vec![NodeOrToken::Token(GreenToken::new(
```
```suggestion
    use NixLanguage::kind_to_raw;
    use NodeOrToken::{Node, Token};
    let pairs: Vec<NodeOrToken<_, _>> = attr_pairs
        .iter()
        .map(move |(k, v)| Node(new_key_value(k.clone(), v.clone())))
        .collect::<Vec<_>>();
    let open_curly_kind = kind_to_raw(TOKEN_CURLY_B_OPEN);
    let close_curly_kind = kind_to_raw(TOKEN_CURLY_B_CLOSE);
    let attr_set_kind = kind_to_raw(NODE_ATTR_SET);
    let whitespace_kind = kind_to_raw(TOKEN_WHITESPACE);
    let mut token_vec = Vec::new();
    token_vec.push(vec![Token(GreenToken::new(
        open_curly_kind,
        "{",
    ))]);
    token_vec.push(vec![Token(GreenToken::new(
        whitespace_kind,
        "\n",
    ))]);
    token_vec.push(pairs);
    token_vec.push(vec![Token(GreenToken::new(
```

zseri commented on src/parser/input_utils.rs:
```diff
@@ -127,8 +113,55 @@ pub fn new_attr_set(attr_pairs: Vec<(GreenNode, GreenNode)>) -> GreenNode {
+impl From<StringLiteral> for GreenNode {
+    fn from(item: StringLiteral) -> Self {
+        let mut node = GreenNodeBuilder::new();
+        let kind: rowan::SyntaxKind = NixLanguage::kind_to_raw(NODE_STRING);
+        node.start_node(kind);
+        let start_string_kind: rowan::SyntaxKind = NixLanguage::kind_to_raw(TOKEN_STRING_START);
+        node.token(start_string_kind, "\"");
+        let string_content: rowan::SyntaxKind = NixLanguage::kind_to_raw(TOKEN_STRING_CONTENT);
+        node.token(string_content, &item.val.to_string());
+        let end_string_kind: rowan::SyntaxKind = NixLanguage::kind_to_raw(TOKEN_STRING_END);
+        node.token(end_string_kind, "\"");
```
```suggestion
        node.start_node(NixLanguage::kind_to_raw(NODE_STRING));
        node.token(NixLanguage::kind_to_raw(TOKEN_STRING_START), "\"");
        node.token(NixLanguage::kind_to_raw(TOKEN_STRING_CONTENT), &item.val.to_string());
        node.token(NixLanguage::kind_to_raw(TOKEN_STRING_END), "\"");
```

zseri commented on src/parser/input_utils.rs:
```diff
@@ -127,8 +113,55 @@ pub fn new_attr_set(attr_pairs: Vec<(GreenNode, GreenNode)>) -> GreenNode {
     GreenNode::new(attr_set_kind, tokens)
 }
 
+pub(crate) struct Key {
+    val: SmlStr,
+}
```
I think a simple newtype struct would be a better fit, unless you intend to add more fields.
```suggestion
pub(crate) struct Key(SmlStr);
```

DieracDelta commented:
@zseri sorry I haven't pulled in your changes yet. I think that in order for this to work in a scalable manner (we'd like to pull in some sort of evaluation mechanism so we can meaningfully make edits and check their correctness), the IR needs to be a bit more flushed out. Going to focus on that for a bit then recheck in the input code.

zseri commented on src/ir/hlir_types.rs:
```diff
@@ -0,0 +1,21 @@
+    String,
+    Function,
+    List(Box<NixType>),
+    AttrSet(Vec<Box<NixType>>),
```
how is this `AttrSet` type decl supposed to work?

zseri commented on src/ir/hlir_types.rs:
```diff
@@ -0,0 +1,21 @@
+    Str(SmlStr),
+}
+
+trait NixValue {}
```
why not model all possible values here (including rather opaque stuff like function calls)? Maybe it would be a good idea to put this into a separate crate.

zseri commented on src/parser/input_utils.rs:
```diff
@@ -83,118 +89,181 @@ pub fn merge_attr_sets(a1: GreenNode, a2: GreenNode) -> GreenNode {
-        let kind: rowan::SyntaxKind = NixLanguage::kind_to_raw(NODE_KEY);
+#[derive(Debug, Clone, Eq, PartialEq, Default)]
+pub(crate) struct NixAttrSet<T: Into<GreenNode> + Clone> {
+    kvs: Vec<NixKeyValue<T>>,
```
why don't we use something like the `indexmap` crate here?

zseri reviewed (commented):
I would suggest splitting the EDSL-like + conversion between that and `rnix`, `rowan` into a separate crate (maybe `rnix-ddhir`), so we can lay out a cleaner interface for it rather sooner than later.

zseri commented on src/parser/input_utils.rs:
```diff
@@ -0,0 +1,269 @@
+use crate::parser::utils::{string_to_node, NixNode};
+use crate::SmlStr;
```
```suggestion
use smol_str::SmolStr;
```

zseri commented on src/parser/input_utils.rs:
```diff
@@ -0,0 +1,269 @@
+
+#[derive(Debug, Clone, Eq, PartialEq, Default)]
+pub(crate) struct NixKey {
+    pub val: SmlStr,
```
```suggestion
    pub val: SmolStr,
```

zseri commented on src/parser/input_utils.rs:
```diff
@@ -0,0 +1,269 @@
+
+#[derive(Debug, Clone, Eq, PartialEq, Default)]
+pub(crate) struct NixStringLiteral {
+    pub val: SmlStr,
```
```suggestion
    pub val: SmolStr,
```

zseri commented on src/parser/input_utils.rs:
```diff
@@ -0,0 +1,269 @@
+//#[derive(Debug, Clone, Eq, PartialEq)]
+//pub(crate) enum SyntaxStructure {
+//KeyValue(Box<SyntaxStructure>, Box<SyntaxStructure>),
+//Key(SmlStr),
+//StringLiteral(SmlStr),
```
```suggestion
//Key(SmolStr),
//StringLiteral(SmolStr),
```

zseri commented on src/parser/input_utils.rs:
```diff
@@ -0,0 +1,269 @@
+//let mut inputs = Vec::new();
+//if let Some(s) = item.url {
+//inputs.push((
+//SyntaxStructure::Key(SmlStr::new_inline("url")).into(),
```
```suggestion
//SyntaxStructure::Key(SmolStr::new_inline("url")).into(),
```

zseri commented on src/parser/input_utils.rs:
```diff
@@ -0,0 +1,269 @@
+//}
+//if let Some(s) = item.is_flake {
+//inputs.push((
+//SyntaxStructure::Key(SmlStr::new_inline("flake")).into(),
```
```suggestion
//SyntaxStructure::Key(SmolStr::new_inline("flake")).into(),
```

zseri commented on src/parser/input_utils_tests.rs:
```diff
@@ -0,0 +1,65 @@
+use crate::parser::input_utils::{merge_attr_sets, new_attr_set, new_key, new_string, wrap_root};
+use crate::parser::utils::{node_to_string, string_to_node};
+use crate::SmlStr;
```
```suggestion
use smol_str::SmolStr;
```

zseri commented on src/parser/input_utils_tests.rs:
```diff
@@ -0,0 +1,65 @@
+#[test]
+pub fn check_new_attr_set() {
+    let attrset = vec![
+        (
+            SyntaxStructure::Key(SmlStr::new_inline("test1")).into(),
+            SyntaxStructure::StringLiteral(SmlStr::new_inline("value1")).into(),
+        ),
+        (
+            SyntaxStructure::Key(SmlStr::new_inline("test2")).into(),
+            SyntaxStructure::StringLiteral(SmlStr::new_inline("value2")).into(),
+        ),
```
```suggestion
        (
            SyntaxStructure::Key(SmolStr::new_inline("test1")).into(),
            SyntaxStructure::StringLiteral(SmolStr::new_inline("value1")).into(),
        ),
        (
            SyntaxStructure::Key(SmolStr::new_inline("test2")).into(),
            SyntaxStructure::StringLiteral(SmolStr::new_inline("value2")).into(),
        ),
```

zseri commented on src/parser/input_utils_tests.rs:
```diff
@@ -0,0 +1,65 @@
+#[test]
+pub fn check_merge_attr_set() {
+    let attrset = vec![
+        (
+            SyntaxStructure::Key(SmlStr::new_inline("test1")).into(),
+            SyntaxStructure::StringLiteral(SmlStr::new_inline("value1")).into(),
+        ),
+        (
+            SyntaxStructure::Key(SmlStr::new_inline("test2")).into(),
+            SyntaxStructure::StringLiteral(SmlStr::new_inline("value2")).into(),
+        ),
```
```suggestion
        (
            SyntaxStructure::Key(SmolStr::new_inline("test1")).into(),
            SyntaxStructure::StringLiteral(SmolStr::new_inline("value1")).into(),
        ),
        (
            SyntaxStructure::Key(SmolStr::new_inline("test2")).into(),
            SyntaxStructure::StringLiteral(SmolStr::new_inline("value2")).into(),
        ),
```

zseri commented on src/user/mod.rs:
```diff
@@ -124,6 +133,8 @@ pub(crate) enum UserPrompt {
     #[display("{0}")]
+    Bool(bool),
+    #[display("{0}")]
     Other(SmlStr),
```
??? maybe rebase necessary ???

zseri commented on src/ir/hlir_types.rs:
```diff
@@ -0,0 +1,21 @@
+use crate::SmlStr;
```
```suggestion
use smol_str::SmolStr;
```

zseri commented on src/ir/hlir_types.rs:
```diff
@@ -0,0 +1,21 @@
+enum NixPrimitive {
+    Bool(bool),
+    Int(i64),
+    Str(SmlStr),
```
```suggestion
    Str(SmolStr),
```

zseri commented on src/parser/input_utils.rs:
```diff
@@ -0,0 +1,269 @@
+        //let mut inputs_gn = Vec::<Box<dyn>>::new()
+        //if let Some(url) = input.url {
+        //inputs_gn.push((
+        //NixKey{val: SmlStr::new_inline("url")}.into(),
```
```suggestion
        //NixKey{val: SmolStr::new_inline("url")}.into(),
```

zseri commented on src/parser/input_utils.rs:
```diff
@@ -0,0 +1,269 @@
+        //let url_gn: GreenNode = url.into();
+        //}
+        //if let Some(is_flake) = input.is_flake {
+        //inputs_gn.push(NixKey{val: SmlStr::new_inline("flake")}.into(), );
```
```suggestion
        //inputs_gn.push(NixKey{val: SmolStr::new_inline("flake")}.into(), );
```

zseri reviewed (changes requested):
necessary because of the removal of the `SmlStr` newtype
